acvm/src/lib.rs

```diff
@@ -122,7 +124,12 @@ pub trait PartialWitnessGenerator {
                                 unassigned_witness.0,
                             )))
                         } else {
-                            self.solve_black_box_function_call(initial_witness, bb_func)
+                            let status = solve_black_box_function(initial_witness, bb_func);
+                            if matches!(status, Err(OpcodeResolutionError::OpcodeNotSolvable(_))) {
+                                self.solve_black_box_function_call(initial_witness, bb_func)
+                            } else {
+                                status
+                            }
```

Comment:
This change is unrelated to supporting keccak opcodes and has significant knock on effects for all backends so shouldn't be included in this PR.

Reply:
I don't think it is a breaking change, what is broken?

Reply:
It's changing the contract between backends and ACVM.

Previously ACVM provided reference methods for executing black box functions to the backends however backends were in charge of solving the black box function and returning the result. After this change this is still the case for all other opcodes but we'll never send keccak256 in particular to the backend.

This split between keccak256 and all the other functions doesn't make sense and we shouldn't split responsibility for black box function execution between the backend and the ACVM. Until we have a rust implementation of all black box functions then we should allow the backend to handle this.

Reply:
So in fact this is not breaking anything.
Solving Keccak in ACVM does make sense, as well as all other 'blackbox'. I will move the solving into ACVM for all but pedersen, which will still be handled by the backend until we have a rust implementation.

Reply:
This change would mean that a lot of code in aztec_backend would become dead code without us actually realizing it. The backend has a match on all black box functions but those code paths would never be executed.

This could lead to confusing bugs between Noir and the backend and it seems hard to communicate to backend implementors.

I'm torn on where this should happen because I don't want aztec_backend to implement keccak itself 😬 I'd probably be fine making a separate PR for this change, just so we can mark it breaking and have a clear CHANGELOG entry. @TomAFrench thoughts?

Reply:
Of course we would remove the dead code as long as it is integrated in ACVM, so I don't see this as an issue.

Reply:
> Of course we would remove the dead code as long as it is integrated in ACVM, so I don't see this as an issue.

This assumes that we are in control of all the backends, but we need to develop as though there are other backends that need to integrate ACVM and they will be confused without accurate documentation & changelog entries.

Reply:
We could land something like https://github.com/noir-lang/acvm/pull/257 before this and then it wouldn't be an issue.